server/application_runner.go

```diff
@@ -0,0 +1,363 @@
+						stage := getActiveStage(issue.Pipeline.StageList)
+						if stage == nil {
+							stage = issue.Pipeline.StageList[len(issue.Pipeline.StageList)-1]
+						}
+						if err != nil {
+							log.Error("failed to get issue by issue id", zap.Int("issueID", externalApproval.IssueID), zap.Error(err))
+							continue
+						}
+						if issue.Status != api.IssueOpen {
+							if _, err := r.server.cancelOldExternalApprovalIfNeeded(ctx, issue, stage, &value, r.P); err != nil {
+								log.Error("failed to cancel external approval", zap.Error(err))
+								continue
+							}
+						}
+
+						status, err := r.P.GetExternalApprovalStatus(ctx, feishu.TokenCtx{
+							AppID:     value.AppID,
+							AppSecret: value.AppSecret,
+							Token:     r.P.Token.Load().(string),
+						}, payload.InstanceCode)
+						if err != nil {
+							log.Error("failed to get approval instance", zap.String("instanceCode", payload.InstanceCode), zap.Error(err))
+							continue
+						}
+
+						if status == "APPROVED" {
+							// double check
+							if stage.ID == payload.StageID && payload.AssigneeID == issue.AssigneeID {
+								// approve stage
+								if err := func() error {
+									var taskIDList []int
+									var tasks []*api.Task
+									for _, task := range stage.TaskList {
+										if task.Status == api.TaskPendingApproval {
+											taskIDList = append(taskIDList, task.ID)
+											tasks = append(tasks, task)
+										}
+									}
+									taskStatusPatch := &api.TaskStatusPatch{
+										IDList:    taskIDList,
+										UpdaterID: externalApproval.ApproverID,
+										Status:    api.TaskPending,
+									}
+									_, err := r.server.store.PatchTaskStatus(ctx, taskStatusPatch)
+									if err != nil {
+										return errors.Wrapf(err, "Failed to update task %q status", taskIDList)
+									}
+									if err := r.server.ActivityManager.BatchCreateTaskStatusUpdateApprovalActivity(ctx, taskStatusPatch, issue, stage, tasks); err != nil {
+										return errors.Wrapf(err, "failed to create task status update activity")
+									}
+									return nil
+								}(); err != nil {
+									log.Error("failed to approve stage", zap.Error(err))
+								}
+
+								if _, err := r.server.store.PatchExternalApproval(ctx, &api.ExternalApprovalPatch{
+									// archive approval instance
+									ID:        externalApproval.ID,
+									RowStatus: api.Archived,
+								}); err != nil {
+									log.Error("failed to archive external apporval", zap.Error(err))
+								}
+							}
+						}
+					default:
+						log.Error("Unknown ExternalApproval.Type", zap.Any("ExternalApproval", externalApproval))
+					}
+				}
+			}()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (s *Server) cancelOldExternalApprovalIfNeeded(ctx context.Context, issue *api.Issue, stage *api.Stage, settingValue *api.SettingAppIMValue, p *feishu.Provider) (*api.ExternalApproval, error) {
```

Comment:
We shouldn't have (s *Server) for stuff in the application runner. And can we avoid passing in the whole server to the runner?